### PR TITLE
More options in DispersionFitter and fit constant loss tangent model

### PR DESCRIPTION
**Pull request by weiliangjin2021**
Addressing https://github.com/flexcompute/tidy3d/issues/1293 and use FastFitter to fit constant loss tangent material.

I'm not sure about the best way to implement https://github.com/flexcompute/tidy3d/issues/1293; currently it's using classmethod to load from complex permittivity and loss tangent, while we still only support wavelength but not frequency.

### CODE REVIEW

**Comment by weiliangjin2021** [comment]
One more comment: now that fitter is fast and stable, should we move material fitter from plugin to tidy3d components, so that we can do `mat = AbstractMedium.constant_loss_tangent_model(eps_real, loss_tangent, frequency_range)` as opposed to `mat = FastDispersionFitter.constant_loss_tangent_model(eps_real, loss_tangent, frequency_range)`?

File: tidy3d/components/medium.py [hunk 1]
```diff
@@ -766,6 +766,21 @@ def nk_model(self, frequency: float) -> Tuple[float, float]:
...
         return self.eps_complex_to_nk(eps_complex)
 
+    def loss_tangent_model(self, frequency: float) -> Tuple[float, float]:
+        """Permittivity and loss tangent as a function of frequency.
+
+        Parameters
+        ----------
+        frequency : float
+            Frequency to evaluate permittivity at (Hz).
+        Returns
```
Comment (tylerflex) [inline tidy3d/components/medium.py:783]: might need a newline

File: tidy3d/plugins/dispersion/fit.py [hunk 2]
```diff
@@ -697,3 +697,63 @@ def from_file(cls, fname: str, **loadtxt_kwargs):
...
         return cls(wvl_um=wvl_um, n_data=n_data, k_data=k_data)
+
+    @classmethod
+    def from_complex_permittivity(
+        cls,
+        wvl_um: ArrayFloat1D,
+        eps_real: ArrayFloat1D,
+        eps_imag: ArrayFloat1D = None,
+        wvl_range: Tuple[Optional[float], Optional[float]] = (None, None),
+    ):
```
Comment (tylerflex) [inline tidy3d/plugins/dispersion/fit.py]: return type? (and below)

File: tidy3d/plugins/dispersion/fit.py [hunk 3]
```diff
@@ -697,3 +697,63 @@ def from_file(cls, fname: str, **loadtxt_kwargs):
...
+        """Loads :class:`DispersionFitter` from wavelength and complex permittivity data
+
+        Parameters
+        ----------
+        wvl_um : ArrayFloat1D
+            Wavelength data in micrometers.
+        eps_real : ArrayFloat1D
+            Real parts of permittivity data
+        eps_imag : Optional[ArrayFloat1D]
+            Imageinary parts of permittivity data; `None` for lossless medium.
```
Comment (tylerflex) [inline tidy3d/plugins/dispersion/fit.py]: Imaginary

File: tidy3d/plugins/dispersion/fit.py [hunk 4]
```diff
@@ -697,3 +697,63 @@ def from_file(cls, fname: str, **loadtxt_kwargs):
...
+
+    @classmethod
+    def from_complex_permittivity(
+        cls,
+        wvl_um: ArrayFloat1D,
+        eps_real: ArrayFloat1D,
+        eps_imag: ArrayFloat1D = None,
+        wvl_range: Tuple[Optional[float], Optional[float]] = (None, None),
+    ):
+        """Loads :class:`DispersionFitter` from wavelength and complex permittivity data
```
Comment (tylerflex) [inline tidy3d/plugins/dispersion/fit.py]: perhaps we should mention here or in the parameters that it's `relative` permittivity? (and for the one below)

**Review by tylerflex — COMMENTED** [review]
Implementation looks good to me!

Don't forget mention this in the changelog btw

**Comment by tylerflex** [comment]
> One more comment: now that fitter is fast and stable, should we move material fitter from plugin to tidy3d components, so that we can do mat = AbstractMedium.constant_loss_tangent_model(eps_real, loss_tangent, frequency_range) as opposed to mat = FastDispersionFitter.constant_loss_tangent_model(eps_real, loss_tangent, frequency_range)?

We can consider that, but we still need to support the plugin for backwards compatibility. So I guess is there a big advantage to moving it?

**Comment by weiliangjin2021** [comment]
> > One more comment: now that fitter is fast and stable, should we move material fitter from plugin to tidy3d components, so that we can do mat = AbstractMedium.constant_loss_tangent_model(eps_real, loss_tangent, frequency_range) as opposed to mat = FastDispersionFitter.constant_loss_tangent_model(eps_real, loss_tangent, frequency_range)?
> 
> We can consider that, but we still need to support the plugin for backwards compatibility. So I guess is there a big advantage to moving it?

We can leave a wrapper in the plugin for compatibility, similar to how we move complex polyslab from plugin to components: https://github.com/flexcompute/tidy3d/blob/14dcf5b953776e6b87624a7423e85bd4e56ea930/tidy3d/plugins/polyslab/polyslab.py#L3-L8

**Comment by tylerflex** [comment]
I see, if you find this would improve usability, feel free to change it. But at the same time, I'm fine with the way things are now too. Up to you

File: tidy3d/plugins/dispersion/fit.py [hunk 5]
```diff
@@ -697,3 +697,63 @@ def from_file(cls, fname: str, **loadtxt_kwargs):
...
+
+        Parameters
+        ----------
+        wvl_um : ArrayFloat1D
+            Wavelength data in micrometers.
+        eps_real : ArrayFloat1D
+            Real parts of permittivity data
+        loss_tangent : Optional[ArrayFloat1D]
+            Loss tangent data, defined as the ratio of imaginary and real parts of permittivity.
+        wvg_range : Tuple[Optional[float], Optional[float]]
```
Comment (dmarek-flex) [inline tidy3d/plugins/dispersion/fit.py]: wvl_range

**Review by dmarek-flex — APPROVED** [review]
Looks good to me, excited to try some new applications.

File: tidy3d/plugins/dispersion/fit.py (hunk 2 above)
Comment (weiliangjin2021) [reply tidy3d/plugins/dispersion/fit.py]: ruff complains here if the return type is "DispersionFitter": F821 Undefined name `DispersionFitter`

Comment (tylerflex) [reply tidy3d/plugins/dispersion/fit.py]: need to `from __future__ import annotations` at the top of file and it should work

Comment (weiliangjin2021) [reply tidy3d/plugins/dispersion/fit.py]: Thanks, it's working now.

**Comment by weiliangjin2021** [comment]
@tylerflex any more comments?

**Comment by tylerflex** [comment]
no, looks good to me

**Review by tylerflex — APPROVED** [review]

**Pull request merged**